zerver/management/commands/export_search.py

```diff
@@ -0,0 +1,119 @@
+from zerver.lib.management import ZulipBaseCommand
+from zerver.models import Message, Recipient, Stream, UserProfile
+
+ignore_keys = [
+    "realm",
+    "rendered_content_version",
+    "sending_client",
+    "search_tsvector",
+]
+
+
+class Command(ZulipBaseCommand):
+    help = """Exports the messages matching certain search terms.
+
+This is most often used for legal compliance.
+"""
+
+    def add_arguments(self, parser: ArgumentParser) -> None:
+        self.add_realm_args(parser, required=True)
+        parser.add_argument(
+            "--output",
+            metavar="<path>",
+            help="File to output JSON results to; it must not exist, unless --force is given",
+            required=True,
+        )
+        parser.add_argument(
+            "--force", action="store_true", help="Overwrite the output file if it exists already"
+        )
+
+        parser.add_argument(
+            "--file",
+            metavar="<path>",
+            help="Read search terms from the named file, one per line",
+        )
+        parser.add_argument(
+            "search_terms",
+            nargs="*",
+            metavar="<search term>",
+            help="Terms to search for in message body or topic",
+        )
+
+    def handle(self, *args: Any, **options: Any) -> None:
+        terms = set()
+        if options["file"]:
+            with open(options["file"], "r") as f:
+                terms.update(f.read().splitlines())
+        terms.update(options["search_terms"])
+
+        if not terms:
+            raise CommandError("One or more search terms are required!")
+
+        if os.path.exists(options["output"]) and not options["force"]:
+            raise CommandError(
+                f"Output path '{options['output']}' already exists; use --force to overwrite"
+            )
+
+        realm = self.get_realm(options)
+        limits = reduce(
+            or_,
+            [Q(content__icontains=term) | Q(subject__icontains=term) for term in terms],
+            Q(),
+        )
+
+        messages_query = Message.objects.filter(limits, realm=realm).order_by("date_sent")
+
+        def format_sender(full_name: str, delivery_email: str) -> str:
+            return str(Address(display_name=full_name, addr_spec=delivery_email))
+
+        @lru_cache(maxsize=None)
+        def format_recipient(recipient_id: int) -> str:
+            recipient = Recipient.objects.get(id=recipient_id)
+
+            if recipient.type == Recipient.STREAM:
+                stream = Stream.objects.values("name").get(id=recipient.type_id)
+                return "#" + stream["name"]
+
+            users = (
+                UserProfile.objects.filter(
+                    subscription__recipient_id=recipient.id,
+                )
+                .order_by("full_name")
+                .values_list("full_name", "delivery_email")
+            )
+
+            return ", ".join([format_sender(e[0], e[1]) for e in users])
```

Comment:
Should we unify this formatting (and perhaps some more of the logic in this script?) with the `export_single_user` codepath?
```
def export_messages_single_user(
    user_profile: UserProfile, *, output_dir: Path, reaction_message_ids: Set[int]
) -> None:
    @lru_cache(maxsize=None)
    def get_recipient(recipient_id: int) -> str:
        recipient = Recipient.objects.get(id=recipient_id)

        if recipient.type == Recipient.STREAM:
            stream = Stream.objects.values("name").get(id=recipient.type_id)
            return stream["name"]

        user_names = (
            UserProfile.objects.filter(
                subscription__recipient_id=recipient.id,
            )
            .order_by("full_name")
            .values_list("full_name", flat=True)
        )

        return ", ".join(user_names)
```

Reply:
That'd be a great follow-up.  There may be a little tension that we want the delivery addresses for the compliance export, but for single-user export the user may not have permissions to see other users' delivery addresses.